### DISCUSSION TITLE
TTY.WTF migrating to Milkdown

### DISCUSSION THREAD

**Discussion by mihailik** [Milkdown/milkdown#1495] [Show and tell]
## TTY.WTF editor

My cutie tiny tool for faux-formatting text using Unicode tricks. [TTY.WTF](https://tty.wtf/%CD%B2%E1%83%A7%CF%81%D2%BD+To+%F0%9F%86%88ourself,/%CF%9F%E1%9B%9E%F0%90%8C%84%F0%90%8C%84%F0%90%8C%95+%E2%88%A6%E1%9B%9C%E1%9B%92%E1%9B%92%E1%9B%82%F0%90%8C%95%CF%9F%F0%90%8C%84%CF%9F) uses CodeMirror 6:

![screenshot of tty.wtf](https://github.com/user-attachments/assets/729376cb-01ee-4f64-bcea-610c99b91984)

Now I'm migrating on Milkdown, and adding a killer new feature: runnable code blocks, early prototype at [md.tty.wtf](https://md.tty.wtf/this+should+be+runnable://%60%60%60JavaScript/Math.random()/%60%60%60//%3E+0.451214345345//%26%23x20;++++++++%5E%5E%5E+imagine+clicking+%22Run%22+and+it+shows//)

![screenshot of md.tty.wtf](https://github.com/user-attachments/assets/d8ceb47f-0d1a-482d-9ec6-1a67ac7d3942)

Wait a minute, that's not just an editor, it becomes...

## Milkdown-powered Notebook

<img src="https://github.com/user-attachments/assets/fa920e2c-30eb-45aa-a496-77b09f919576" align="right" width="300">

One app to bring together formatted content, scripts, data visualisations and bind them in one document.

* Embed JSON data
* Or call HTTP with fetch
* Run all sorts of slicing/dicing
* Throw it into a table, chart, THREE.js scene
* Allow user inputs inside document

Imagine the power of Excel/VBA but not evil and fit for the web.

## What's now

Milkdown is largely integrated, and it's  𝒜 𝒲 𝐸 𝒮 𝒪 𝑀 𝐸 !!!

For runnable code, I need:

* Upgrade the **code block** UX with **[Run]** button
* Handle actual **execution**
* Display the **results** for text/JSON at least

At first just for JavaScript code.

## In future?

TypeScript, obviosly
Python via Brython
Ag-Grid for data output
Input UI elements
Safely access non-public APIs — via OAuth, or Basic auth, or whatever else

Keen to hear any thoughts, suggestions, advice!

And again, thanks **a lot** for this extremely cool project @Saul-Mirone

**Comment by mihailik** [comment]
## I've got script execution POC!!

See a Markdown document with a button to run script block: [md.tty.wtf/...](https://md.tty.wtf/Try+selecting+code+below,+and+clicking+%60run%60://%60%60%60JavaScript/fetch(/++'https:%2F%2Fplc.directory%2Fexport%3Fcount=1000%26after='+%2B/++new+Date(Date.now()-10000).toISOString())/++.then(x=%3Ex.text())/++.then(x=%3E/++++x.split('%5Cn').length.toLocaleString()+%2B/++++'+BlueSky+signups+in+the+last+10+sec')/%60%60%60//) 

![screenshot with fetch](https://github.com/user-attachments/assets/0aef0644-2e2c-4346-a6af-ca007a823169)

Now this is very hacky: the button is injected via CSS ::before and DOM events.

I'd love to integrate a runner feature with Milkdown's code-block properly. Could you give me some directions @Saul-Mirone please?

1. Can I inject my custom DOM above and below `codemirror-host`, around [this line in packages/components/src/**code-block/view/component.ts** L201](https://github.com/Milkdown/milkdown/blob/main/packages/components/src/code-block/view/component.ts#L201)?
2. Can I add custom keyboard shortcut (maybe Cmd-Enter) to trigger my JS execution code?

Do I need a private fork of Milkdown for that, or can I achieve that with customisation of the official library?

Many thanks again!

**Comment by mihailik** [comment]
Or alternative approach: inject execution UX at CodeMirror side.

Following [CodeMirror Panel Example](https://codemirror.net/examples/panel/) I can attach a panel with results/errors/progress at the bottom of the code block without involving Milkdown.

Update: managed to handle it this CodeMirror panel way!

![image](https://github.com/user-attachments/assets/743366a4-ed37-4d07-bd5c-98a44ae701b0)

However it creates weird scrolling issues  when results are taking large space.

Maybe it better belongs to the code-block DOM after all?

Any advice much appreciated ❤️

**Reply by mihailik** [reply]
Having played with CodeMirror-powered code blocks in Milkdown more, I'm thinking it's a dead end.

They are just not sufficiently integrated in the editor. Selection is off, styles off, copy-pasting is weird, adding adornments too hard.

For the complexity I need, the code-block logic will have to be re-implemented natively on ProseMirror. That's also a chance to base highlights and rich code interactions on VSCode tooling, i.e. TypeScript language server and so on.

**Comment by trancethehuman** [comment]
The design looks awesome

**Reply by mihailik** [reply]
Thank you!!

I've got a fair bit ahead with a custom codeblock implementation, that pulls the result text into the document as a readonly island.

![image](https://github.com/user-attachments/assets/b1fd6bcd-7725-45fa-b012-9966d8e0e8c7)

[tty.wtf/...](https://tty.wtf/This+script+fetches+recent+1+second+worth+of+BlueSky+account+operations,+then+extracts+human-readable+data://%60%60%60js/var+howManySeconds+=+1//fetch('https:%2F%2Fplc.directory%2Fexport%3Fafter='+%2B+new+Date(Date.now()+-+1000+*+howManySeconds).toISOString())/++.then(x=%3Ex.text()).then(text=%3E+Object.fromEntries(/++++text.trim().split('%5Cn').map(chunk+=%3E/++++++chunk+%26%26+JSON.parse(chunk)).filter(Boolean).map(json+=%3E+%5Bjson.did,+%7B/++endpoint:+json.operation%3F.services%3F.atproto_pds%3F.endpoint,/++handle:+json.operation%3F.alsoKnownAs%3F.%5B0%5D%3F.replace('at:%2F%2F',''),/++createdAt:+json.createdAt/%7D%5D)))/%60%60%60//)

This way navigation across the Milkdown document is seamless, you can also copy-paste script results with no second thought.

Still more needed to restore syntax highlights and results formatting, but the example above should give a taste of the potential.

**Comment by mihailik** [comment]
## Massive update since!

Try yourself: [TTY.WTF/Examples+with+code+blocks...](https://tty.wtf/%23%23+Examples+with+code+blocks://**JS**+is+executed+and+highlighted.+The+results+come+up+in+the+bar+below.//%60%60%60JavaScript/%2F**+@param+%7Bnumber%7D+n+*%2F/function+fibonacci(n)+%7B+return+!n+%3F+0+:+n+%3C=+2+%3F+1+:+fibonacci(n+-+1)+%2B+fibonacci(n+-+2);+%7D/%60%60%60//**TS**+is+highlighted+only,+not+executed.//%60%60%60TypeScript/function+fibonacciTS(n:+number)+%7B/++return+!n+%3F+0+:+n+%3C=+2+%3F+1+:+fibonacci(n+-+1)+%2B+fibonacci(n+-+2);/%7D/%60%60%60//Code+blocks+share+execution+context,+and+can+refer+to+each+other's+variables://%60%60%60/(%7B/++fibonacci:+typeof+fibonacci+===+'function'+%3F+fibonacci(12)+:+'no',/++fibonacciTS:+typeof+fibonacciTS+===+'function'+%3F+fibonacciTS(12)+:+'no',/++alert:+typeof+alert/%7D)/%60%60%60//**JSON**+is+highlighted+only,+not+executed.//%60%60%60json/%5B/++%7B+%22id%22:+%22apple%22,+%22name%22:+%22Apply%22%7D,/++%7B+%22id%22:+%22pear%22,+name:+12e1+%7D/%5D/%60%60%60//All+of+them,+JavaScript,+TypeScript+and+JSON+are+highlighted,+provide+tooltips+and+recognize+errors.//Other+languages+are+left+as+plain+text+(but+plugin-based+providers+can+add+support!).//%60%60%60SQL/select+*+from+ARTICLES/%60%60%60//)

![image](https://github.com/user-attachments/assets/57168611-94bf-46a7-889b-2453ba46cc1c)

* Syntax highlight, error squigglies and tooltips are implemented using language services from TypeScript.
* TypeScript library is loaded asynchronously, not blocking initial load.
* All browser typings are pulled at runtime (lib.d.ts).
* JS code is evaluated asynchronously in an isolated IFRAME, and results displayed below script.
* Result area is effectively read-only, you can only delete it together with whole code-block.

There's been really a lot of work done, redone and refactored.

## Ongoing focus:

* Typing completions for JS/TS
* Execution improvements, including transforming the code before running:
  * import statements
  * console.log outputs
* Supporting DOM execution output, by embedding hosting IFRAME into Milkdown document
* Adding HTTP language support, executing by making a request

**Reply by mihailik** [reply]
Here I've had to remove the Milkdown own code-block handling, and replace with more complex one.

The key difference is that Milkdown sees code-block as a simple block-level node, with contained **text** and a **lang** attribute:

<table><tr><td>

**code_block** - block-level, content: `text*`, lang: "JavaScript"

Math.sin(1)

</td></tr></table>

A more complex structure is needed to support results in-document, compare with the above for the same Markdown:

<table><tr><td>

**code_block** - a block-level element, with rigid content: `code_block_backtick_language  code_block_script  code_block_execution_state`

<table><tr><td>

**code_block_backtick_language** - block-level, content: `text*`

JavaScript

</td></tr></table>

<table><tr><td>

**code_block_script** - block-level, content: `text*`

Math.sin(1)

</td></tr></table>

<table><tr><td>

**code_block_execution_state** - block-level, content: `text*`

0.8414709848078965

</td></tr></table>

</td></tr></table>

This [ProseMirror schema](https://github.com/oyin-bo/rest/blob/3f8edebe3c799f162f4371143a62208514bfd342/src/markdown/code-block/schema.js) allows smooth navigation and selection of code blocks in Milkdown document.

Interactive selection/cursor can freely move in/out of execution results sub-block. This is important.

When used for real purpose, those results would contain relevant and important calculation data. Users will want to copy-paste out these to other places. That's why it's important to have these results as smoothly embedded in the document.

**Comment by mihailik** [comment]
Another massive leap in progress!

[TTY.wtf/BSki+identity+operations...](https://tty.wtf/BSki+identity+operations//%60%60%60url/plc.directory%2Fexport%3Fafter=2024-10-15/%60%60%60////Latest+100+posts+from+firehose://%60%60%60/(async+()+=%3E+%7B/++const+coldsky+=+await+import('https:%2F%2Funpkg.com%2Fcoldsky');/++const+msgs+=+%5B%5D;/++for+await(const+chunk+of+coldsky.firehose())+%7B/++++for(const+m+of+chunk.messages)+%7B/++++++if(m.$type+===+'app.bsky.feed.post')/++++++++msgs.push(m);/+++++%7D/++++if+(msgs.length+%3E=+100)+return+msgs;/++%7D/%7D)()/%60%60%60//)

![17290833954618235580105441804322](https://github.com/user-attachments/assets/b31b84c7-3e17-4052-82f0-f75b452abaa2)

* JavaScript, JSON, TypeScript semantic highlight
* Syntax and semantics error (red squiggles) TS/JS/json
* Tooltips - including browser API docs
* Auto completion as you type
* Execution for JavaScript
* Fetching for HTTP language: either raw URL or full HTTP request with headers, method etc.
* Auto-detecting tabular response data, formatting into a grid

Near future focus will be on look-n-feel boost, and supporting Python and SQL code blocks.

Not just syntax highlight, execution environment for Python and SQL.

Data fetched via HTTP can be munched by JS, sliced by Python and further queries by SQL **in the same document!**

[See in action](https://tty.wtf/BSki+identity+operations//%60%60%60url/plc.directory%2Fexport%3Fafter=2024-10-15/%60%60%60////Latest+100+posts+from+firehose://%60%60%60/(async+()+=%3E+%7B/++const+coldsky+=+await+import('https:%2F%2Funpkg.com%2Fcoldsky');/++const+msgs+=+%5B%5D;/++for+await(const+chunk+of+coldsky.firehose())+%7B/++++for(const+m+of+chunk.messages)+%7B/++++++if(m.$type+===+'app.bsky.feed.post')/++++++++msgs.push(m);/+++++%7D/++++if+(msgs.length+%3E=+100)+return+msgs;/++%7D/%7D)()/%60%60%60//)

**Reply by mihailik** [reply]
Switching from CodeMirror to native ProseMirror took a great deal of effort, but it's so much smoother.

Cursor movement, selection, change transactions — all just speak the same language 🤌

I've figured out the interactivity too, like clicking on ℹ️ expands error details, basically updating the ProseMirror document and decorations.

**Reply by mihailik** [reply]
A big chunk of work was outside the editor: a model for safe execution.

I'm creating a hidden IFRAME for that, on a subdomain like 123abc-ifrwrk.tty.wtf — which leads to full isolation by the browser.

That IFRAME is talking to the editor via postMessage.

That'll get even more funky with Python execution!

**Reply by mihailik** [reply]
But I do want to allow visual outputs in future (like for code snippets involving THREE.js)

So that hidden IFRAME may need to become unhidden one day.

This is such a rabbit hole of a project 🐇
